src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.cs

```diff
@@ -21,6 +21,10 @@ public sealed partial class JsonSerializerOptions
 
         private readonly ConcurrentDictionary<Type, JsonClassInfo> _classes = new ConcurrentDictionary<Type, JsonClassInfo>();
 
+        // Simple LRU cache for the public (de)serialize entry points that avoid some lookups in _classes.
+        // Although this may be written by multiple threads, 'volatile' was not added since any local affinity is fine.
```

Comment:
It looks like the test `ConstructorTests.MultipleTypes` would catch any issues here since it deserializes across multiple Types and multiple threads, but I'll add another specific test that makes it clear.